### PR TITLE
[ENH] Refactor `plot_matrix`

### PR DESCRIPTION
**Pull request by NicolasGensollen**
This PR proposes to:

- refactor the code of the function `plot_matrix` which is too long with a hard-to-follow control flow.
- add more tests for `plot_matrix`
- refactor old test function for `plot_matrix` into multiple smaller tests
- improve the docstrings of the functions in the file `matrix_plotting.py`.
- adjust a little bit the space for the colorbar to avoid having the colorbar too close to the matrix (see a few examples like [this one](https://nilearn.github.io/auto_examples/03_connectivity/plot_signal_extraction.html#sphx-glr-auto-examples-03-connectivity-plot-signal-extraction-py))

### CODE REVIEW

File: nilearn/plotting/matrix_plotting.py [hunk 1]
```diff
@@ -37,6 +40,151 @@ def fit_axes(ax):
...
+    if axes is not None and figure is not None:
+        raise ValueError(
+            "Parameters figure and axes cannot be specified "
+            "together. You gave 'figure=%s, axes=%s'" % (figure, axes)
+        )
+    if figure is not None:
+        if isinstance(figure, plt.Figure):
+            fig = figure
+        else:
+            fig = plt.figure(figsize=figure)
```
Comment (bthirion) [inline nilearn/plotting/matrix_plotting.py:66]: A bit weird to code figsize in figure...

File: nilearn/plotting/matrix_plotting.py [hunk 2]
```diff
@@ -37,6 +40,151 @@ def fit_axes(ax):
...
+        axes = plt.gca()
+        own_fig = True
+    else:
+        if axes is None:
+            fig, axes = plt.subplots(1, 1, figsize=(7, 5))
```
Comment (bthirion) [inline nilearn/plotting/matrix_plotting.py:71]: Why is (7, 5) a good figsize ?

**Review by bthirion — COMMENTED** [review]
Lgtm overal, thx.

File: nilearn/plotting/matrix_plotting.py (hunk 1 above)
Comment (NicolasGensollen) [reply nilearn/plotting/matrix_plotting.py:66]: I agree, but the docstring of `plot_matrix` says you can do it:

```
figure : :class:`matplotlib.figure.Figure`, figsize :obj:`tuple`,\
        or None, optional
        Sets the figure used. This argument can be either an existing
        figure, or a pair (width, height) that gives the size of a
        newly-created figure.
```

If you think we shouldn't allow this, we'd have to deprecate it.

File: nilearn/plotting/matrix_plotting.py (hunk 2 above)
Comment (NicolasGensollen) [reply nilearn/plotting/matrix_plotting.py:71]: I don't know why this was chosen as a default. 
To make it a little less hidden in the code, I can add a global constant DEFAULT_FIGSIZE at the beginning of the file and add a note in the docstring saying that if no figure is provided, a new one will be created with this default size.

File: nilearn/plotting/matrix_plotting.py (hunk 1 above)
Comment (bthirion) [reply nilearn/plotting/matrix_plotting.py:66]: This is a bit weird, but let's keep that outside of this PR ---given the existing docstrings.

File: nilearn/plotting/matrix_plotting.py [hunk 3]
```diff
@@ -37,6 +40,151 @@ def fit_axes(ax):
...
+        axes.set_xticklabels(labels, size='x-small')
+        for label in axes.get_xticklabels():
+            label.set_ha('right')
+            label.set_rotation(50)
+        axes.set_yticks(np.arange(len(labels)))
+        axes.set_yticklabels(labels, size='x-small')
+        for label in axes.get_yticklabels():
+            label.set_ha('right')
+            label.set_va('top')
+            label.set_rotation(10)
```
Comment (bthirion) [inline nilearn/plotting/matrix_plotting.py:162]: I'm a bit worried about all these hardcoded paarmeters. See also l.156.

File: nilearn/plotting/matrix_plotting.py [hunk 4]
```diff
@@ -98,127 +258,38 @@ def plot_matrix(mat, title=None, labels=None, figure=None, axes=None,
...
                              rect=((0, 0, .95, 1) if colorbar
                                    else (0, 0, 1, 1)))
-
     if colorbar:
         cax, kw = make_axes(axes, location='right', fraction=0.05, shrink=0.8,
                             pad=.0)
         fig.colorbar(mappable=display, cax=cax)
         # make some room
-        fig.subplots_adjust(right=0.8)
+        fig.subplots_adjust(right=0.78)
```
Comment (bthirion) [inline nilearn/plotting/matrix_plotting.py:292]: Another hard-coded parameter...

**Review by bthirion — APPROVED** [review]
LGTM, thx !

**Review by bthirion — COMMENTED** [review]
LGTM overall, thx !

**Comment by NicolasGensollen** [comment]
Hi @tsalo 
Do you think you can make a review sometime this week such that we can merge this for 0.9?
Thanks a lot!

**Pull request merged**